tezos/context-api/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+    InMem,
+}
+
+// Must be in sync with ffi_config.ml
+#[derive(Clone, Serialize, Deserialize, Debug)]
+pub struct TezosContextTezEdgeStorageConfiguration {
+    pub backend: ContextKvStoreConfiguration,
+    pub ipc_socket_path: Option<String>,
+}
+
+// Must be in sync with ffi_config.ml
+#[derive(Clone, Serialize, Deserialize, Debug)]
+pub enum TezosContextStorageConfiguration {
+    IrminOnly(TezosContextIrminStorageConfiguration),
+    TezEdgeOnly(TezosContextTezEdgeStorageConfiguration),
+    Both(
+        TezosContextIrminStorageConfiguration,
+        TezosContextTezEdgeStorageConfiguration,
+    ),
+}
+
+impl TezosContextStorageConfiguration {
+    /// Used to produce a configuration for the readonly protocol runners from the configuration used in the main protocol runner.
+    ///
+    /// If only Irmin is enabled, the resulting configuration is the same.
+    /// If only TezEdge is enabled, the resulting configuration switches the backend to the readonly IPC implementation.
+    /// If both Irmin and TezEdge are enabled, an only-Irmin configuration is enabled.
+    pub fn readonly(&self) -> Self {
+        match self {
+            TezosContextStorageConfiguration::IrminOnly(_) => self.clone(),
+            TezosContextStorageConfiguration::TezEdgeOnly(tezedge) => {
+                TezosContextStorageConfiguration::TezEdgeOnly(
+                    TezosContextTezEdgeStorageConfiguration {
+                        backend: ContextKvStoreConfiguration::ReadOnlyIpc,
+                        ..tezedge.clone()
+                    },
+                )
+            }
+            TezosContextStorageConfiguration::Both(_irmin, tezedge) => {
+                TezosContextStorageConfiguration::TezEdgeOnly(
+                    TezosContextTezEdgeStorageConfiguration {
+                        backend: ContextKvStoreConfiguration::ReadOnlyIpc,
+                        ..tezedge.clone()
+                    },
+                )
+            }
+        }
+    }
+
+    pub fn get_ipc_socket_path(&self) -> Option<String> {
+        match self {
+            TezosContextStorageConfiguration::IrminOnly(_) => None,
+            TezosContextStorageConfiguration::TezEdgeOnly(tezedge) => {
+                tezedge.ipc_socket_path.clone()
+            }
+            TezosContextStorageConfiguration::Both(_irmin, tezedge) => {
+                tezedge.ipc_socket_path.clone()
+            }
+        }
+    }
+
+    pub fn tezedge_is_enabled(&self) -> bool {
+        match self {
+            TezosContextStorageConfiguration::IrminOnly(_) => false,
+            TezosContextStorageConfiguration::TezEdgeOnly(_) => true,
+            TezosContextStorageConfiguration::Both(_, _) => true,
+        }
+    }
+}
+
+// Must be in sync with ffi_config.ml
+#[derive(Serialize, Deserialize, Debug)]
+pub struct TezosContextConfiguration {
+    pub storage: TezosContextStorageConfiguration,
+    pub genesis: GenesisChain,
+    pub protocol_overrides: ProtocolOverrides,
```

Comment:
@tizoc 
nice cleanup :),
technically, `ProtocolOverrides` and `enable_testchain` are not related to context, they are just used with `apply_block` function, so maybe "sometime" we should move it from `TezosContextConfiguration` and `context-api`

Reply:
@bkontur yes. Also, while this solves the shell -> libtezos dependency, something that is still pending is the full split between our context storage and everything else so that it can be packaged as an opam package.

Reply:
Btw, if you spot anything else please point it out. For now I am leaving everything as-is, because the focus was on properly splitting things, but now may be a good time to start caring about this kind of stuff too while I work on the IPC messages.